thread_sync.c

```diff
@@ -369,6 +397,7 @@ do_mutex_lock(VALUE self, int interruptible_p)
                 // unlocked by another thread while sleeping
                 if (!mutex->fiber) {
                     mutex->fiber = fiber;
+                    mutex->thread = th->self;
```

Comment:
These probably both need write barriers now that we mark them (and all the similar assignments), but I can test more thoroughly with wbcheck to be sure